### PR TITLE
[GLUTEN-2961][VL] Integrate with upstream Velox

### PR DESCRIPTION
**Pull request by rui-mo**
## What changes were proposed in this pull request?

Make the needed changes to integrate with upstream Velox, and start to depend on a new Velox branch [update](https://github.com/oap-project/velox/tree/update), which shall be rebased frequently to keep updated with Velox upstream.

Changes mainly including:

1. Change to use Velox provided Arrow 13.0.
2. Integrate with Velox decimal type and date type refactor.
3. Integrate with Velox memory module updates.
4. Integrate with Velox parquet writer updates.
5. Function integration and fixes.
6. Solve Velox library linking issues, and fix Gluten CI issues.

Lacks:
- ORC support
-  Performance gap on TPC-H/DS
- GHA docker image update

## How was this patch tested?

Verified the functionality on TPC-DS, TPC-H and Spark unit tests.

### CODE REVIEW

File: pom.diff [hunk 1]
```diff
@@ -0,0 +1,12 @@
+diff --git a/backends-velox/pom.xml b/backends-velox/pom.xml
```
Comment (rui-mo) [inline pom.diff:1]: To be removed.

File: cpp/velox/operators/functions/RegistrationAllFunctions.cc [hunk 2]
```diff
@@ -39,17 +51,18 @@ void registerFunctionOverwrite() {
   facebook::velox::registerFunction<RoundFunction, double, double, int32_t>({"round"});
   facebook::velox::registerFunction<RoundFunction, float, float, int32_t>({"round"});
 }
+} // anonymous namespace
```
Comment (rui-mo) [inline cpp/velox/operators/functions/RegistrationAllFunctions.cc:54]: // namespace

File: cpp/velox/utils/VeloxArrowUtils.cc [hunk 3]
```diff
@@ -58,7 +58,7 @@ arrow::Status MyMemoryPool::Allocate(int64_t size, int64_t alignment, uint8_t**
   if (bytes_allocated() + size > capacity_) {
     return arrow::Status::OutOfMemory("malloc of size ", size, " failed");
   }
-  RETURN_NOT_OK(pool_->Allocate(size, alignment, out));
+  RETURN_NOT_OK(pool_->Allocate(size, out));
```
Comment (rui-mo) [inline cpp/velox/utils/VeloxArrowUtils.cc:61]: Is this change needed?

**Comment by zhouyuan** [comment]
CC: @zuochunwei

File: cpp/velox/utils/VeloxArrowUtils.cc (hunk 3 above)
Comment (marin-ma) [reply cpp/velox/utils/VeloxArrowUtils.cc:61]: Yes, because arrow version is downgraded, and the API changed.

**Comment by JkSelf** [comment]
@rui-mo https://github.com/facebookincubator/velox/pull/5437 has merged. Can we also pick this PR in velox?

File: backends-velox/src/test/scala/io/glutenproject/execution/TestOperator.scala [hunk 4]
```diff
@@ -458,11 +458,11 @@ class TestOperator extends WholeStageTransformerSuite {
       .select($"DecimalCol".cast(DecimalType(38, 33)))
       .select(col("DecimalCol"))
       .agg(avg($"DecimalCol"))
-    assert(result.collect()(0).get(0).toString.equals("0.0345678900000000000000000000000000000"))
+    assert((result.collect()(0).get(0).toString.toDouble - d).abs < 0.00000000001)
```
Comment (JkSelf) [inline backends-velox/src/test/scala/io/glutenproject/execution/TestOperator.scala:461]: @rui-mo It is also due to https://github.com/facebookincubator/velox/pull/6051#issuecomment-1731028215? If yes, can we also add the comment?

File: .github/workflows/velox_be.yml [hunk 5]
```diff
@@ -55,17 +55,14 @@ jobs:
...
-          ./build_arrow.sh --arrow_home=/opt/arrow --enable_ep_cache=ON && \
           cd /opt/gluten/ep/build-velox/src && \
           ./get_velox.sh --velox_home=/opt/velox && \
           ./build_velox.sh --velox_home=/opt/velox --enable_ep_cache=ON'
       - name: Build Gluten CPP library
         run: |
           docker exec ubuntu2004-test-$GITHUB_RUN_ID bash -c '
           cd /opt/gluten/cpp && \
-          ./compile.sh --build_velox_backend=ON --arrow_home=/opt/arrow --velox_home=/opt/velox --build_tests=ON --build_examples=ON --build_benchmarks=ON'
+          ./compile.sh --build_velox_backend=ON --velox_home=/opt/velox --arrow_home=/opt/velox/_build/release/third_party/arrow_ep --build_tests=ON --build_examples=ON --build_benchmarks=ON'
```
Comment (zhztheplayer) [inline .github/workflows/velox_be.yml:65]: So /opt/arrow will never be used after this change, right?

File: gluten-ut/spark32/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala [hunk 6]
```diff
@@ -348,6 +358,14 @@ class VeloxTestSettings extends BackendTestSettings {
...
       "columns (nested data)")
     .exclude("SPARK-31116: Select nested schema with case insensitive mode")
+    // ReaderFactory is not registered for format orc.
+    .exclude("SPARK-15474 Write and read back non-empty schema with empty dataframe - orc")
+    .exclude("SPARK-23271 empty RDD when saved should write a metadata only file - orc")
+    .exclude("SPARK-22146 read files containing special characters using orc")
+    .exclude("Do not use cache on overwrite")
+    .exclude("Do not use cache on append")
+    .exclude("File source v2: support partition pruning")
+    .exclude("File source v2: support passing data filters to FileScan without partitionFilters")
```
Comment (zhztheplayer) [inline gluten-ut/spark32/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala:368]: Was there already a plan to re-enable these excluded tests?

Comment (zhztheplayer) [reply gluten-ut/spark32/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala:368]: OK I see there was reason put in `excludeAll`

File: gluten-ut/spark32/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala [hunk 7]
```diff
@@ -385,11 +403,13 @@ class VeloxTestSettings extends BackendTestSettings {
   enableSuite[GlutenOrcPartitionDiscoverySuite]
     .exclude("read partitioned table - normal case")
     .exclude("read partitioned table - with nulls")
+    .excludeAll("Blocked by ORC Velox upstream not ready")
```
Comment (zhztheplayer) [inline gluten-ut/spark32/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala:406]: Why not just comment the lines out? I think that will be less confusing to developers.

File: backends-velox/src/test/scala/io/glutenproject/execution/TestOperator.scala (hunk 4 above)
Comment (rui-mo) [reply backends-velox/src/test/scala/io/glutenproject/execution/TestOperator.scala:463]: Updated.

File: gluten-core/src/main/scala/io/glutenproject/expression/DateTimeExpressionsTransformer.scala [hunk 8]
```diff
@@ -76,8 +76,8 @@ case class DateDiffTransformer(
       // In CH backend, datediff params are ('day', startDate, endDate).
       Lists.newArrayList(ExpressionBuilder.makeStringLiteral("day"), startDateNode, endDateNode)
     } else {
-      // In the others, datediff params are (startDate, endDate).
-      Lists.newArrayList(startDateNode, endDateNode)
+      // In the others, datediff params are (endDate, startDate).
+      Lists.newArrayList(endDateNode, startDateNode)
```
Comment (zhztheplayer) [inline gluten-core/src/main/scala/io/glutenproject/expression/DateTimeExpressionsTransformer.scala:80]: Is this because new version of Velox changed the arg order of datediff?

File: cpp/velox/utils/VeloxArrowUtils.cc (hunk 3 above)
Comment (rui-mo) [reply cpp/velox/utils/VeloxArrowUtils.cc]: The arrow version is upgraded from arrow 12.0 to 13.0, so I assume alignment could be kept.

File: gluten-ut/spark32/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala (hunk 7 above)
Comment (marin-ma) [reply gluten-ut/spark32/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala:406]: There are lots of lines to comment out, meanwhile need to remove the "imports" to pass style check because those classes are not used after comment. We can simply remove this line after ORC is supported.

Comment (zhztheplayer) [reply gluten-ut/spark32/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala:406]: e.g.

```scala
// Blocked by ORC Velox upstream not ready
//  enableSuite[GlutenOrcPartitionDiscoverySuite]
//    .exclude("read partitioned table - normal case")
//    .exclude("read partitioned table - with nulls")
```

File: .github/workflows/velox_be.yml (hunk 5 above)
Comment (rui-mo) [reply .github/workflows/velox_be.yml:65]: Yes, it is not in use.

File: gluten-core/src/main/scala/io/glutenproject/expression/DateTimeExpressionsTransformer.scala (hunk 8 above)
Comment (rui-mo) [reply gluten-core/src/main/scala/io/glutenproject/expression/DateTimeExpressionsTransformer.scala:80]: Yes, but the new one has aligned parameter order as Spark.
> case class DateDiff(endDate: Expression, startDate: Expression)

File: gluten-ut/spark32/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala (hunk 7 above)
Comment (zhztheplayer) [reply gluten-ut/spark32/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala:406]: > There are lots of lines to comment out, meanwhile need to remove the "imports" to pass style check because those classes are not used after comment. We can simply remove this line after ORC is supported.

Sounds fair to me. Let me try untangle this part of code a bit directly on the PR.

File: .github/workflows/velox_be.yml [hunk 9]
```diff
@@ -82,11 +79,11 @@ jobs:
...
-      # Cpp micro benchmarks will use generated files from unit test in backends-velox module.
-      - name: Run micro benchmarks
-        run: |
-          docker exec ubuntu2004-test-$GITHUB_RUN_ID bash -c 'cd /opt/gluten/cpp/build/velox/benchmarks && \
-          ./generic_benchmark --with-shuffle --threads 1 --iterations 1'
+#       Cpp micro benchmarks will use generated files from unit test in backends-velox module.
+#      - name: Run micro benchmarks
+#        run: |
+#          docker exec ubuntu2004-test-$GITHUB_RUN_ID bash -c 'cd /opt/gluten/cpp/build/velox/benchmarks && \
+#          ./generic_benchmark --with-shuffle --threads 1 --iterations 1'
```
Comment (zhouyuan) [inline .github/workflows/velox_be.yml:86]: benchmark is commented out, is this necessary?

File: .github/workflows/velox_be.yml [hunk 10]
```diff
@@ -394,6 +375,7 @@ jobs:
         run: |
           docker exec -i static-build-test-$GITHUB_RUN_ID bash -c '
           source /env.sh && \
+          sudo yum -y install patch && \
```
Comment (zhouyuan) [inline .github/workflows/velox_be.yml:378]: move this to velox dep?

**Review by zhouyuan — COMMENTED** [review]
"Some compression codecs are not supported in Velox parquet write" -> can you please list the missing codec here?

File: cpp/velox/substrait/SubstraitParser.cc [hunk 11]
```diff
@@ -317,11 +316,13 @@ std::string SubstraitParser::findVeloxFunction(
 std::string SubstraitParser::mapToVeloxFunction(const std::string& substraitFunction, bool isDecimal) {
   auto it = substraitVeloxFunctionMap_.find(substraitFunction);
   if (isDecimal) {
-    if (substraitFunction == "add" || substraitFunction == "subtract" || substraitFunction == "multiply" ||
-        substraitFunction == "divide" || substraitFunction == "avg" || substraitFunction == "avg_merge" ||
-        substraitFunction == "sum" || substraitFunction == "sum_merge" || substraitFunction == "round") {
+    if (substraitFunction == "round" || substraitFunction == "lt" || substraitFunction == "lte" ||
```
Comment (yma11) [inline cpp/velox/substrait/SubstraitParser.cc:319]: Does it mean that functions like "devide" won't be supported after switching to upstream velox?

Comment (rui-mo) [reply cpp/velox/substrait/SubstraitParser.cc:319]: No, it's just a name mapping. For decimal type, decimal_divide will used.

File: .github/workflows/velox_be.yml (hunk 9 above)
Comment (rui-mo) [reply .github/workflows/velox_be.yml:86]: This benchmark generates ORC file with Arrow ORC file writer, but ORC is not compiled in Velox Arrow. We need to enable that first.

https://github.com/oap-project/gluten/pull/3341/files#diff-f1db35cb133919c45a122798963e00bebe0c957fcb8a552bd80951961fc0123fR28-R30

**Comment by rui-mo** [comment]
> "Some compression codecs are not supported in Velox parquet write" -> can you please list the missing codec here?

@zhouyuan I picked lz4 support to Velox branch, and they are all supported now.

**Comment by rui-mo** [comment]
> @rui-mo [facebookincubator/velox#5437](https://github.com/facebookincubator/velox/pull/5437) has merged. Can we also pick this PR in velox?

@JkSelf It is included now.

File: .github/workflows/velox_be.yml (hunk 10 above)
Comment (rui-mo) [reply .github/workflows/velox_be.yml:378]: I find `setup_linux` is called only for vcpkg compilation, so it does not work to add this command in setup-centos7.sh.
https://github.com/oap-project/gluten/blob/main/ep/build-velox/src/build_velox.sh#L76-L78

**Comment by marin-ma** [comment]
We still need to maintain GenericBenchmark for debug purpose. Let me try to extract the ORC reader in micro benchmark by adding a compile option which is initially disabled.

File: gluten-ut/spark32/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala (hunk 7 above)
Comment (zhztheplayer) [reply gluten-ut/spark32/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala:406]: https://github.com/oap-project/gluten/pull/3402

**Review by zhouyuan — APPROVED** [review]
👍  big change!

**Comment by rui-mo** [comment]
Just merged this PR, and the remaining issues can be solved directly on the main branch from now on.
Below are the backups for today's Gluten and Velox:
https://github.com/oap-project/gluten/commits/2023_10_19
https://github.com/oap-project/velox/tree/2023_10_19
Thank you all for the efforts on this change!

**Comment by zhouyuan** [comment]
note: there are several todo
- fix the performance gap
- fix ORC related tests
- improve GHA CI

**Comment by Yohahaha** [comment]
Great work!

**Comment by Yohahaha** [comment]
@zhouyuan 
> fix the performance gap

could you share more about this?

**Comment by zhouyuan** [comment]
@Yohahaha Yes, on TPC-H/DS we find ~20% perf regression, the issue is mostly on hashjoin and hashagg. We already identified some root cause on Hash Agg (some tunings on param required). Will have some follow commits as this patch is quite big and requires rebase efforts from Rui.   

Here's one report on TPCH SF2000
https://github.com/oap-project/gluten/pull/3341#issuecomment-1769752364

thanks, -yuan

**Comment by Yohahaha** [comment]
> @Yohahaha Yes, on TPC-H/DS we find ~20% perf regression, the issue is mostly on hashjoin and hashagg. We already identified some root cause on Hash Agg (some tunings on param required). Will have some follow commits as this patch is quite big and requires rebase efforts from Rui.
> 
> Here's one report on TPCH SF2000 [#3341 (comment)](https://github.com/oap-project/gluten/pull/3341#issuecomment-1769752364)
> 
> thanks, -yuan

We found ~1x perf regression in TPCDS sf1000.

**Comment by zhouyuan** [comment]
> > @Yohahaha Yes, on TPC-H/DS we find ~20% perf regression, the issue is mostly on hashjoin and hashagg. We already identified some root cause on Hash Agg (some tunings on param required). Will have some follow commits as this patch is quite big and requires rebase efforts from Rui.
> > Here's one report on TPCH SF2000 [#3341 (comment)](https://github.com/oap-project/gluten/pull/3341#issuecomment-1769752364)
> > thanks, -yuan
> 
> We found ~1x perf regression in TPCDS sf1000.

@Yohahaha can you please check https://github.com/oap-project/gluten/pull/3439 can help to bring some performance back?

thanks, -yuan

**Comment by Yohahaha** [comment]
@zhouyuan 
after 3439, OOM throwed, seems peek memory is higher.
